File: ecdsa-verify-fsharp/ECDSAVerify/ECDSAVerify.fs
namespace ECDSAVerify

open System
open System.Numerics

type ECDSAVerify = 
    static member ModularInverse(a: BigInteger, n: BigInteger) = 
        if n = BigInteger.Zero || n = BigInteger.One then
            None
        else
            let rec loop t newt r newr =
                if newr = BigInteger.Zero then
                    if r > BigInteger.One then
                        None
                    else
                        if t < BigInteger.Zero then
                            Some (t + n)
                        else
                            Some t
                else
                    let quotient = r / newr
                    loop newt (t - quotient * newt) newr (r - quotient * newr)
            loop BigInteger.Zero BigInteger.One n a

File: ecdsa-verify-csharp/ECDSAVerify/src/ECDSAVerify.cs
using System;
using System.Linq;
using System.Numerics;

namespace ECDSAVerify
{
    public static class ECDSAVerify
    {
        public static BigInteger? ModularInverse(BigInteger a, BigInteger n)
        {
            if (n == BigInteger.Zero || n == BigInteger.One)
            {
                return null;
            }

            BigInteger t = BigInteger.Zero;
            BigInteger newt = BigInteger.One;
            BigInteger r = n;
            BigInteger newr = a;
            while (newr != BigInteger.Zero)
            {
                BigInteger quotient = r / newr;
                (t, newt) = (newt, t - quotient * newt);
                (r, newr) = (newr, r - quotient * newr);
            }

            if (r > BigInteger.One)
            {
                return null;
            }

            if (t < BigInteger.Zero)
            {
                t = t + n;
            }

            return t;
        }
    }
}